relay-system/src/runtime.rs

```diff
@@ -3,75 +3,83 @@ use tokio::task::JoinHandle;
 
 use crate::statsd::SystemCounters;
 
-/// Spawns a new asynchronous task, returning a [`JoinHandle`] for it.
+/// Spawns an instrumented task with an automatically generated [`TaskId`].
 ///
-/// This is in instrumented spawn variant of Tokio's [`tokio::spawn`].
+/// Returns a [`JoinHandle`].
 #[macro_export]
 macro_rules! spawn {
     ($future:expr) => {{
-        static _TASK_ID: ::std::sync::OnceLock<$crate::TaskId> = ::std::sync::OnceLock::new();
-        let task_id = _TASK_ID.get_or_init(|| (*::std::panic::Location::caller()).into());
-        $crate::_spawn_inner(task_id, $future)
+        static _PARTS: ::std::sync::OnceLock<(String, String, String)> =
+            ::std::sync::OnceLock::new();
+        let (id, file, line) = _PARTS.get_or_init(|| {
+            let caller = *::std::panic::Location::caller();
+            let id = format!("{}:{}", caller.file(), caller.line());
+            (id, caller.file().to_owned(), caller.line().to_string())
+        });
+        $crate::spawn((id.as_str(), file.as_str(), line.as_str()), $future)
```

Comment:
I would create the `TaskId` here explicitly and not through the `From/Into` calls, something like `TaskId::from_internal` and hide it with `#[doc(hidden)]`, I think this is a bit less magic than the tuple.